lib/online_order.rb

```diff
@@ -0,0 +1,103 @@
+require 'awesome_print'
+require 'pry'
+require 'csv'
+require_relative 'order'
+require_relative 'customer'
+
+
+module Grocery
+
+# Custom ArgumentError because it's fun, could also be done at the point of raising the ArgumentError using syntax ArgumentError.new("Custom Error Message")
+  class CartError < ArgumentError
+    def initialize(msg="Error: Items cannot be added to this order based on its current status")
+      super
+    end
+  end
+
+  class OnlineOrder < Order
+    attr_reader :customer, :status
+    @@all_online_orders = []
+    @@customers = []
+
+    def initialize(id, products, customer_id, status = :pending)
+      #Use the itialize attributes from the parent class Order
+      super(id, products)
+      @customer = customer_id
+      @status = status.to_sym
+    end # Initialize
+
+    def total
+      unless @products.length == 0
+        return super + 10
+      end
+    end
+
+    def add_product(product_name, product_price)
+      if @status == :paid || @status == :pending
+        super
+      else
+        begin
+          raise CartError
+        rescue
+        end
+      end
+    end
+
+    def self.customers
+      return @@customers
+    end
+
+    def self.all
+      if @@all_online_orders.length == 0
+        self.populate
+      end
+        return @@all_online_orders
+    end
+
+    def self.populate
+      CSV.open('support/online_orders.csv', 'r', headers: true, header_converters: :symbol).each do |row|
+        products = {}
+        id = row[:id].to_i
+        customer_id = row[:customer_id].to_i
+        status = row[:status].to_sym
+        row[:products].split(';').each do |item|
+          name, price = item.split(':')
+          products[name] = price.to_f
+        end
+        @@all_online_orders << index = self.new(id, products, customer_id, status)
+        @@customers << Grocery::Customer.find(customer_id)
+      end
+      return @@all_online_orders
+    end # OnlineOrder.all
+
+    def self.find(id)
+      populate
+      if id > @@all_online_orders.length
+        raise Grocery::FindError.new
```

Comment:
This assumes that the IDs in the CSV file are sequential. In this project this ends up being true, but as we get into working with more complex databases this won't always be the case.

If the IDs weren't sequential, you could rewrite this code as:

```ruby
def self.find(id)
  populate
  @@all_online_orders.each do |order|
    return order if order.id == id
  end
  raise Grocery::FindError.new
end
```